client/app/react-components/QueryEditor.js

```diff
@@ -0,0 +1,205 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { map } from 'lodash';
+
+import AceEditor from 'react-ace';
+import ace from 'brace';
+import toastr from 'angular-toastr';
+
+import 'brace/ext/language_tools';
+import 'brace/mode/json';
+import 'brace/mode/python';
+import 'brace/mode/sql';
+import 'brace/theme/textmate';
+
+const langTools = ace.acequire('ace/ext/language_tools');
+const snippetsModule = ace.acequire('ace/snippets');
+
+// By default Ace will try to load snippet files for the different modes and fail.
+// We don't need them, so we use these placeholders until we define our own.
+function defineDummySnippets(mode) {
+  ace.define(`ace/snippets/${mode}`, ['require', 'exports', 'module'], (require, exports) => {
+    exports.snippetText = '';
+    exports.scope = mode;
+  });
+}
+
+defineDummySnippets('python');
+defineDummySnippets('sql');
+defineDummySnippets('json');
+
+function buildKeywordsFromSchema(schema) {
+  const keywords = {};
+  schema.forEach((table) => {
+    keywords[table.name] = 'Table';
+    table.columns.forEach((c) => {
+      keywords[c] = 'Column';
+      keywords[`${table.name}.${c}`] = 'Column';
+    });
+  });
+  return map(keywords, (v, k) => ({
+    name: k,
+    value: k,
+    score: 0,
+    meta: v,
+  }));
+}
+
+export default class QueryEditor extends React.Component {
+  static propTypes = {
+    queryText: PropTypes.string.isRequired,
+    schema: PropTypes.array, // eslint-disable-line react/no-unused-prop-types
+    syntax: PropTypes.string,
+    dataSources: PropTypes.array,
+    dataSource: PropTypes.object,
+    isQueryOwner: PropTypes.bool.isRequired,
+    updateDataSource: PropTypes.func.isRequired,
+    canExecuteQuery: PropTypes.func.isRequired,
+    executeQuery: PropTypes.func.isRequired,
+    saveQuery: PropTypes.func.isRequired,
+    updateQuery: PropTypes.func.isRequired,
+    listenForResize: PropTypes.func.isRequired, // eslint-disable-line react/no-unused-prop-types
+  }
+
+  static defaultProps = {
+    syntax: 'sql',
+    autocompleteQuery: false,
+    schema: null,
+    dataSource: { options: { doc: '' } },
+    dataSources: [],
+  }
+
+  static getDerivedStateFromProps(nextProps, prevState) {
+    if (!nextProps.schema) {
+      return { ...prevState, keywords: [], autocompleteQuery: false };
+    } else if (nextProps.schema !== prevState.schema) {
+      return {
+        ...prevState,
+        schema: nextProps.schema,
+        keywords: buildKeywordsFromSchema(nextProps.schema),
+        autocompleteQuery: (nextProps.schema.reduce((totalLength, table) =>
+          totalLength + table.columns.length, 0) <= 5000),
+      };
+    }
+    return prevState;
```

Comment:
`getDerivedStateFromProps` should return `null` if state doesn't need to be updated.

Reply:
Thanks for that link. That's some really interesting reading.